### PR TITLE
Argument type hints

### PR DESCRIPTION
**Pull request by lushoBarlett**
Takes function arguments' type hints into account for type inference. As a sidenote, I consider that we may benefit on a way to test for program failures as well.

### CODE REVIEW

**Comment by SebastianMestre** [comment]
> ... we may benefit on a way to test for program failures as well.

That will be trivial once we start elevating type errors to the caller instead of asserting on them. For now, that's pretty much impossible (or at least really hard, by using multiprocessing and stuff).

**Review by SebastianMestre — CHANGES_REQUESTED** [review]
This looks good to me, except for a tiny tiny slip in metacheck.

File: src/metacheck.cpp [hunk 1]
```diff
@@ -47,10 +47,11 @@ void metacheck(TypedAST::Identifier* ast, TypeChecker& tc) {
 void metacheck(TypedAST::FunctionLiteral* ast, TypeChecker& tc) {
 	assign_meta_type(ast->m_meta_type, tc.meta_value(), tc);
 
-	int arg_count = ast->m_args.size();
-	for (int i = 0; i < arg_count; ++i)
-		// TODO: metacheck type hints
-		assign_meta_type(ast->m_args[i].m_meta_type, tc.meta_value(), tc);
+	for (auto& arg : ast->m_args) {
+		if (arg.m_type_hint)
+			metacheck(arg.m_type_hint, tc);
```
[the remark this answers is not in the document]
Comment (lushoBarlett) [reply src/metacheck.cpp:52]: woops

File: src/metacheck.cpp [hunk 2]
```diff
@@ -48,8 +48,10 @@ void metacheck(TypedAST::FunctionLiteral* ast, TypeChecker& tc) {
 	assign_meta_type(ast->m_meta_type, tc.meta_value(), tc);
 
 	for (auto& arg : ast->m_args) {
-		if (arg.m_type_hint)
+		if (arg.m_type_hint) {
 			metacheck(arg.m_type_hint, tc);
+			assign_meta_type(arg.m_type_hint, tc.meta_monotype(), tc);
```
Comment (SebastianMestre) [inline src/metacheck.cpp:53]: should be

```c++
assign_meta_type(arg.m_type_hint->m_meta_type, tc.meta_monotype(), tc);
```

**Review by SebastianMestre — CHANGES_REQUESTED** [review]

File: src/metacheck.cpp (hunk 2 above)
Comment (lushoBarlett) [reply src/metacheck.cpp:53]: This is what happens when I try and do it from github instead of turning on my other computer

**Review by SebastianMestre — APPROVED** [review]
Looks good to me

**Pull request merged**